File: src/apply-middleware.ts
import { Middleware, Producer } from "./types";
import { entries } from "./utils/entries";

/**
 * Creates a producer enhancer that applies middleware to every dispatcher
 * function in the producer.
 * @param middlewares The middleware to apply.
 * @returns The producer enhancer.
 */
export function applyMiddleware<T extends Producer<any, any>>(...middlewares: Middleware[]): (producer: T) => T;
export function applyMiddleware(...middlewares: Middleware[]) {
	return (producer: Producer<any, any>) => {
		const dispatchers = producer.getDispatchers();

		let currentDispatcher: string | undefined;

		const resolveDispatcher = () => {
			assert(currentDispatcher, "Cannot resolve dispatcher outside of middleware");
			return currentDispatcher;
		};

		for (const [name, dispatcher] of entries<string, Callback>(dispatchers)) {
			let dispatch = dispatcher;

			for (const i of $range(middlewares.size() - 1, 0, -1)) {
				const middleware = middlewares[i];
				dispatch = middleware(dispatch, resolveDispatcher, producer);
			}

			const startDispatch = dispatch;

			dispatch = (...args: unknown[]) => {
				currentDispatcher = name;
				return startDispatch(...args);
			};

			dispatchers[name] = dispatch;
			producer[name] = dispatch;
		}

		return producer;
	};
}

File: src/applyMiddleware.lua
local types = require(script.Parent.types)

type Enhancer = (producer: types.Producer) -> types.Producer

local function applyMiddleware(...: types.Middleware): Enhancer
	local arguments = table.pack(...)

	return function(producer: types.Producer)
		local middlewares: { [number]: (...any) -> any, n: number } = table.clone(arguments)
		local dispatchers: { [string]: (...any) -> any } = producer:getDispatchers()

		for index, middleware in ipairs(middlewares) do
			middlewares[index] = middleware(producer)
		end

		for name, dispatcher in dispatchers do
			for index = middlewares.n, 1, -1 do
				dispatcher = middlewares[index](dispatcher, name)
			end

			(producer :: any)[name] = dispatcher
			dispatchers[name] = dispatcher
		end

		return producer
	end
end

return applyMiddleware
